specs/Standards/Wallets/InjectedWallets.md

```diff
@@ -0,0 +1,369 @@
+
+## Specification
+
+Injected wallets are browser extensions that implement the `Wallet` API (see below) on the `window` object. To avoid namespace collisions and easily detect when they're available, wallets will mount under their own key within `window.near` (e.g. `window.near.sender`).
+
+At the core of a wallet are [`signTransaction`](#signtransaction) and [`signTransactions`](#signtransactions). These methods, when given a [`Transaction`](https://nomicon.io/RuntimeSpec/Transactions) instance from [`near-api-js`](https://github.com/near/near-api-js), will prompt the user to sign with a key pair previously imported (with the assumption it has [`FullAccess`](https://nomicon.io/DataStructures/AccessKey) permission).
+
+In most cases, a dApp will need a reference to an account and associated public key to construct a [`Transaction`](https://nomicon.io/RuntimeSpec/Transactions). The [`connect`](#connect) method helps solve this issue by prompting the user to select one or more accounts they would like to make visible to the dApp. When at least one account is visible, the wallet considers the dApp [`connected`](#connected) and they can access a list of [`accounts`](#accounts) containing an `accountId` and `publicKey`.
+
+For dApps that often sign gas-only transactions, [`FunctionCall`](https://nomicon.io/DataStructures/AccessKey#accesskeypermissionfunctioncall) access keys can be added/deleted for one or more accounts using the [`signIn`](#signin) and [`signOut`](#signout) methods. While this functionality could be achieved with [`signTransactions`](#signtransactions), it suggests a direct intention that a user wishes to sign in/out of a dApp's smart contract.
+
+### Wallet API
+
+Below is the entire API for injected wallets. It makes use of `near-api-js` to enable interoperability with dApps that will already use it for constructing transactions and communicating with RPC endpoints.
+
+```ts
+import { transactions, utils } from "near-api-js";
+
+interface Account {
+  accountId: string;
+  publicKey: utils.PublicKey;
+}
+
+interface Network {
+  networkId: string;
+  nodeUrl: string;
+}
+
+interface SignInParams {
+  permission: transactions.FunctionCallPermission;
+  accounts: Array<Account>;
+}
+
+interface SignOutParams {
+  accounts: Array<Account>;
+}
+
+interface SignTransactionParams {
+  transaction: transactions.Transaction;
+}
+
+interface SignTransactionsParams {
+  transactions: Array<transactions.Transaction>;
+}
+
+interface Events {
+  accountsChanged: { accounts: Array<Account> };
+  networkChanged: { network: Network };
+}
+
+type Unsubscribe = () => void;
+
+interface Wallet {
+  id: string;
+  connected: boolean;
+  network: Network;
+  accounts: Array<Account>;
+  
+  connect(): Promise<Array<Account>>;
```

Comment:
Should there not be some way for the dapp to indicate which network they are using? Right now, a `testnet` dapp might open the wallet with `connect()` and the user might select accounts on the wrong network.

Reply:
Perhaps the `Account` instances should have that information:

```typescript
interface Account {
  accountId: string;
  publicKey: utils.PublicKey;
  networkId: string;
}
```
I don't think we should implicitly rely on `accountId`s having that information due to implicit accounts and TLAs

Reply:
This would not be necessary actually since the `Wallet` interface has the `network` property and the dapp would only call `connect` on wallets with a `networkId` relevant to it.

Reply:
I don't really understand? For instance, our wallet allows both testnet and mainnet connections- the same goes for most other wallets as well I think?

Where is the network set by the dapp, basically? Since they would need to indicate that somewhere.

Right now this Wallet Interface is the entire wallet object set on the window object... Are we expected to set multiple, one for each network? That would not be a great solution imo.

Reply:
I think the `Wallet` instance is a lot simpler and more extensible when it targets one `nodeUrl` and therefore `networkId`. It's analogous to a provider in this way.

There is no functionality specified here that interacts with the node but if it's added by the wallet or specified in future we'd have to keep solving the problem of which node we are talking to. Wallets can set multiple `Wallet` instances if they support multiple networks at the same time but generally injected wallets support 1 network at a time and will remove one of the `Wallet` instances when the user toggles between them.

Reply:
> but generally injected wallets support 1 network at a time and will remove one of the Wallet instances when the user toggles between them.

How would this work? Where is this "toggle" happening?

This is what I'm trying to get at- we need to have some form of communication from the dapp side as to which wallet instance is required (i.e. which network the dapp is targeting).

> I think the Wallet instance is a lot simpler and more extensible when it targets one nodeUrl and therefore networkId. It's analogous to a provider in this way.

I can agree, that once the network we are using for interaction is established- having the wallet act as a single network "provider" could be useful and extensible. But first we need a way to communicate the required network, where is that done?

Reply:
> How would this work? Where is this "toggle" happening?

This happens on the wallet UI, e.g. metamask has a dropdown on top right

> we need a way to communicate the required network, where is that done?

The dapp would iterate through the wallets in window until it finds a wallet that matches the `networkId` it's looking to interact with

Reply:
> This happens on the wallet UI, e.g. metamask has a dropdown on top right

Are we not talking about a stage that is long before that? We are on a dapp site- the user may not have interacted with their wallet for a while now, or maybe recently had interacted with a testnet site. But now a mainnet dapp is going to ignore their wallet completely because it just wasn't recently "set" to the right network.

I feel like this is going to be a pain going forward and create more UX issues between dapps and users. And the way to avoid this would just be a simple parameter which the dapp can send to the injected wallet object. Again, most wallets are multi-tenancy, they can handle both cases with ease. Its really not that hard to set an internal parameter for the network and create a specific "provider" of wallet functionality.

Another issue- what if the user was signed into a testnet dapp, but then opened their wallet to check out a mainnet account for some reason, send some funds etc. And then close the wallet again, with the full intention of still interacting with the current dapp where they left off. With the current spec, this is going to create a bit of an ambiguous situation in the dapp now. These are the kinds of headaches that could easily be solved by just being more deterministic, as much as possible, from both sides.

Reply:
I agree with idea that API should be as clear as possible. Favoring being deterministic instead of "convenience" feels like the right choice for me.

So having `window.near.wallet.connect({ networkId: 'mainnet' })` instead of `window.near.wallet.connect()`
And to be clear we could also introduce a `window.near.wallet.supportedNetworks()`

There are multiple ways how this API may look and even a solution which still keeps `Wallet` tied to a single network may be used:
`const testnetWallet = window.near.wallet.getWallet('testnet')`

Reply:
@lostpebble @PinkiNice Thanks for the discussion, we appreciate the insights.  The idea currently discussed here was to drive the network and accounts list _from the wallet to the dApp_, using events (`networkChanged` and `accountsChanged`).  It does create an issue if the dApp does not change to reflect the new network or accounts list where the wallet UI chosen network may not match the dApp's chosen network.

I've updated the spec to eliminate the event emitted for `networkChanged` from the wallet to the dApp context, and instead inverted the control so that the dApp requests the wallet for a particular network ID during `connect()`, and there is an implicit guarantee that the wallet instance on the window that we reference from then on will never reference any other network.

Thanks for the discussion points!  If I have missed the mark on this at all, please follow up on #408 :)